### PR TITLE
Adding min and max (and slice)

### PR DESCRIPTION
**Pull request by pascallouisperez**
tl;dr see `rFoldNumbers` and then how this generic helper is used to write `min`, `max`, `sum`. Adding a `slice` func to create slices inline, which helps with testing.

Adding helper `slice` to create slices, and adding `min` and `max` which works over a wide range of numerical arguments: trying to allow any possible ways in which numbers could be passed, list of numbers, slices of numbers, slices of slices of numbers, a combination of all of the above. For that, introducing a way to 'fold' numerical arguments which is a just a way to iterate over all numerical arguments in a generic way.

Changing `sum` to be based over the same scaffolding which `min` and `max` are based on.

### CODE REVIEW

File: expressions.go [hunk 1]
```diff
@@ -538,6 +513,130 @@ func rFirstOf(args *fnArgs) (Value, error) {
 	return vUndefined, nil
 }
 
+type foldNumbers interface {
+	update(value *Number)
+	result() Value
+}
+
+func rFoldNumbers(f foldNumbers, args *fnArgs, min int) (Value, error) {
```
Comment (lgonzalez-silen) [inline expressions.go]: minArgs is clearer

File: expressions.go [hunk 2]
```diff
@@ -538,6 +513,130 @@ func rFirstOf(args *fnArgs) (Value, error) {
...
+	}
+	return f.result(), nil
+}
+
+type sumFolder struct {
+	sum *Number
+}
+
+func (f *sumFolder) update(value *Number) {
+	f.sum = f.sum.Plus(value)
```
Comment (lgonzalez-silen) [inline expressions.go]: perhaps allow start with a nil sum like in min/maxFolder? then you wouldn't need to init to vZero

File: expressions.go [hunk 3]
```diff
@@ -538,6 +513,130 @@ func rFirstOf(args *fnArgs) (Value, error) {
...
+		if _, ok := arg.Type().(*UndefinedType); !ok {
+			if elementType == nil {
+				elementType = arg.Type()
+			} else if nElementType, ok := elementType.(*NumberType); ok {
+				if nArgType, ok := arg.Type().(*NumberType); ok {
+					if nElementType.scale < nArgType.scale {
+						elementType = nArgType
+					}
+				} else {
+					return nil, fmt.Errorf("cannot mix incompatible types %s and %s in slice", elementType, arg.Type())
```
[the remark this answers is not in the document]
Comment (pascallouisperez) [reply expressions.go]: Because we want to allow `number[0]` coming before `number[1]`, we need to handle that way:
- `number[1]` is not assignable to `number[0]`, but we can extend to inferring that the slice type should be `number[1]`
- but then, the else else handles cases where we mix numerical and non-numerical (or two non-numerical)
I've added a test which covers the other branch.

File: expressions.go (hunk 1 above)
Comment (pascallouisperez) [reply expressions.go]: changing, thx

File: expressions.go (hunk 2 above)
Comment (pascallouisperez) [reply expressions.go]: if you allow `nil`, then this code would be `if f.sum == nil { f.sum = vZero }`, right? I felt it'd be easier to initialize to `vZero`, plus we want a sum of an empty list to be 0, see https://github.com/helloeave/worksheets/blob/master/computed_by_test.go#L915
